feat: Added IMAP for automated 2FA

Skribb11es opened the pull request:
Need to make some documentation for the usage, but as is, it has been tested on Gmail and works perfectly. This essentially allows a user to input their IMAP credentials for their email account, and then upon a 2FA code being requested by the script, it will then log into the user's email account and wait for the email to come in from Riot. At this point, it will strip the code from the email and immediately log out of the account. It then uses this code for verification, similar to a manual input, but fully automated. Exclusions and logging have been included as part of this addition, along with forcefully closing an account's thread if the specified IMAP information is incorrect, as this may lead to the user's account being locked down if it was to loop.

Basic docs - 
Open the config.yaml file and within specify the needed information for your IMAP account according to your email provider. For gmail this information can be found here - https://support.google.com/mail/answer/7126229 (do note if you are going to use Gmail with this, google requires that you have 2FA enabled on your google account, and that you make a dedicated "App Password". The then-created app password will be used instead of your google account's password.)


"imapusername" - The username that is specified by your provider to be used when logging into IMAP. For Gmail this is your full email address.

"imappassword" - The password that is specified by your provider to be used when logging into IMAP. For Gmail this is the "App Password" that you created.

"imapserver" - The specified server that your provider is looking for you to connect to. EX - imap.gmail.com

Once the information has been specified within the config (per account), you can then simply run the farmer, and it should all be taken care of.

auroraisluna commented:
Interesting feature, looks good to me. 

I suggest not including these credentials in the config file as it's an optional thing, and users may consider this required. (because they can't read the wiki :D)

auroraisluna commented on config/config.yaml:
```diff
@@ -2,6 +2,12 @@ accounts:
   accountname:
     username: "username"
     password: "password"
+    imapusername: "" # Leave this blank if you do not want to use automatic 2FA, or do not use 2FA
```
Referring about these

auroraisluna commented on config/config.yaml:
```diff
@@ -2,6 +2,12 @@ accounts:
-    password: "password"
\ No newline at end of file
+    password: "password"
+    imapusername: "" # Leave this blank if you do not want to use automatic 2FA, or do not use 2FA
```
and here

Skribb11es commented:
Figured it would be more obvious that it exists hence the addition of the comment, however, it can be done if need be, really figured that it would be a better idea to keep it in the config as default and have a comment stating that it was optional so users know that it's there without needing to look at the wiki, but don't need to use it if they don't want too.

auroraisluna commented:
I see your point but keep in mind 70% of the users that open the config, have managed to mess it up. Let's not add additional points of failure, if possible. 

That being said, you should raise an issue (or add it your self if thats fine with you) to also update the `confighelper.html` to include this feature.

Skribb11es commented:
Removed the IMAP configuration options from the default config, and also updated the confighelper to include the ability to make configs with IMAP integration. (Also fixed an issue with a redirect it had)

LeagueOfPoro commented on src/FarmThread.py:
```diff
@@ -61,8 +62,13 @@ def run(self):
+            if InvalidIMAPCredentialsException:
+                self.log.error(f"IMAP login failed for {self.account}")
+                self.stats.updateStatus(self.account, "[red]IMAP LOGIN FAILED")
+                self.stats.accountData[self.account]["failedLoginCounter"] = 3
```
See PR #90

Skribb11es replied:
Will do, but still need to include a full stop in some form or fashion due to 2fa "errors" more specifically accounts being locked down in the event that too many 2fa codes are requested but not used.

LeagueOfPoro commented on src/IMAP.py:
```diff
@@ -0,0 +1,52 @@
+from threading import *
+import re
+ 
+class Idler(object):
```
Please match the class and file name

DjordyKoert commented on src/Config.py:
```diff
@@ -1,93 +1,98 @@
-            return Path("config/config.yaml")
-        
-        return configPath
+import yaml
```
Line terminator has most likely been changed https://stackoverflow.com/questions/19593909/git-diff-sees-whole-file-as-changed-when-its-not

Skribb11es commented on src/Browser.py:
```diff
@@ -61,15 +65,34 @@ def login(self, username: str, password: str, refreshLock) -> bool:
+                        M.select("INBOX")
+                        idler = IMAP.Idler(M)
+                        idler.start()
+                        self.stats.updateStatus(self.account, f"[green]WAITING FOR 2FA")
```
It is shown, and it has consistently throughout my testing since its addition. If proof is need it can be supplied, but I promise it works as intended.

Skribb11es commented on config/confighelper.html:
```diff
@@ -1,111 +1,157 @@
+  <div class="main">
+    <h1>Capsule Farmer Evolved<br><span>Config Generator</span></h1>
+    <div class="card">
+      <h3>Informations</h3>
```
This can be done, will either supply one to Alepouna since they love writing docs, or add one myself. Former is probably better as I'm pretty shit at writing docs.

Skribb11es commented on config/config.yaml:
```diff
@@ -4,4 +4,4 @@ accounts:
     password: "password"
   anotheraccountname:
     username: "username"
-    password: "password"
```
No change was intended, will be undone.
